Add initial porting guidelines (#95)
AlexsLemonade/OpenScPCA-nf: porting-modules.md

If multiple output files are created by a module (e.g., a table of results and an R object with more detailed output), the same general format should be followed, but with additional entries in each channel element: `[sample_id, project_id, output_files_1, output_files_2, ...]`.

Where possible, individual output files should contain the `SCPCS` sample id, `SCPCL` library id, or `SCPCP` project id, as appropriate, to facilitate searching and filtering.

## Module processes

### Containers

Each process should run in a container, usually the container defined in `OpenScPCA-analysis` for the module that the process is a part of, which will be available on the [AWS Public ECR](https://gallery.ecr.aws/openscpca/).
The container should be defined with a version tag to ensure that the container used is consistent across all runs of the workflow (though `latest` is acceptable during development).

**Question:** Do we want to have a single config file with all container definitions? Defining the container within a module is deprecated, but I am not sure that we really need variables for this since each module will only have a limited number of processes.

### Process granularity

> There are no hard and fast rules about how grandular each process should be, as we want to balance the workflow complexity with flexibility and runtime efficiency.

Reviewer: though i do very much like this word as a portmanteau of grand and granular, and i otherwise endorse it's use

In general, if one step in the module's analysis is particularly long-running or resource-intensive, it may be better to break that step into a separate process, as we can then assign it higher resource requirements while leaving the other steps running in low-resource nodes.
In addition, if the intermediate files from a process are going to be useful for other analyses, it is better to have a separate process that emits those files as output.
If, however, the intermediate files are only useful within the context of the module, and each step is relatively fast, it may be better to have a single process that includes multiple steps where only the final output is emitted.

### Process resources

The default resources for each process are 4 GB of memory and 1 CPU.
Any additional requirements should be defined with `label` directives in the process definition.
Available labels are defined in `config/process_base.config`, and separate labels are used for memory and CPU requirements.

> For example, to request 16 GB of memory and 4 CPUs, the process definition would include the following:

Reviewer: ```suggestion

For example, to request 16 GB of memory and 4 CPUs, the process definition would include the following:
```

```groovy
process my_process {
    label 'mem_16'
    label 'cpus_4'
    ...
}
```

> If an instance of a process fails, the memory requirements are automatically increased on the second and third attempts, but the general goal should be to have each process run to completion for the majority of samples with the assigned resources.

Reviewer: ```suggestion
If an instance of a process fails, the memory requirements are automatically increased on the second and third attempts, but the general goal should be for each process successfully complete the majority of samples with the assigned resources.
```

### Module input (take)

In general, module workflows should `take:` as input the `sample_ch` channel that is defined in `OpenScPCA-nf` default workflow.
Each element of this channel has the following structure: `[sample_id, project_id, file(sample_dir)]`

The final element is a file/path object, and could be passed directly to a Nextflow process to stage all data files for a sample.
However, this is not recommended, as most processes will only require a subset of files, such as only the processed `AnnData` files (and not raw files or `SingleCellExperiment` files).
Instead, the files that will be required for each sample should be selected using the [`Utils.getLibraryFiles()` function](#the-utilsgetlibraryfiles-function) or similar methods.

#### The `Utils.getLibraryFiles()` function

The `Utils.getLibraryFiles()` function is designed to create a list of files that are relevant to a particular sample that can be passed as input to a process for proper data staging.

The function takes the following arguments:
- `sample_dir` – The path to the sample directory
- `format:` – The format of the files to be selected (`sce` or `anndata`)

> - `process_level:` – The processing level of the files to be selected (`raw`, `filtered` or `processed`)

Author: The colons are actually an important part of the syntax here, so I would like to leave them.

> - `process_level:` – The processing level of the files to be selected (`raw`, `filtered` or `processed`)

Reviewer: Ah, I see! At a quick glance I thought it was a typo with list explanations
